matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/pushers/DefaultAddPusherTaskTest.kt

```diff
@@ -0,0 +1,178 @@
+
+private val A_JSON_PUSHER = JsonPusher(
+        pushKey = "push-key",
+        kind = "http",
+        appId = "m.email",
+        appDisplayName = "Element",
+        deviceDisplayName = null,
+        profileTag = "",
+        lang = "en-GB",
+        data = JsonPusherData(brand = "Element")
+)
+
+class DefaultAddPusherTaskTest {
+
+    private val pushersAPI = FakePushersAPI()
+    private val monarchy = FakeMonarchy()
+
+    private val addPusherTask = DefaultAddPusherTask(
+            pushersAPI = pushersAPI,
+            monarchy = monarchy.instance,
+            requestExecutor = FakeRequestExecutor(),
+            globalErrorReceiver = FakeGlobalErrorReceiver()
+    )
+
+    @Test
+    fun `given no persisted pusher when adding Pusher then updates api and inserts result with Registered state`() {
+        monarchy.givenWhereReturns<PusherEntity>(result = null)
+
+        runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+
+        pushersAPI.verifySetPusher(A_JSON_PUSHER)
+        monarchy.verifyInsertOrUpdate<PusherEntity> {
+            withArg { actual ->
+                actual.state shouldBeEqualTo PusherState.REGISTERED
+            }
+        }
+    }
+
+    @Test
+    fun `given a persisted pusher when adding Pusher then updates api and mutates persisted result with Registered state`() {
+        val realmResult = PusherEntity(appDisplayName = null)
+        monarchy.givenWhereReturns(result = realmResult)
+
+        runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+
+        pushersAPI.verifySetPusher(A_JSON_PUSHER)
+
+        realmResult.appDisplayName shouldBeEqualTo A_JSON_PUSHER.appDisplayName
+        realmResult.state shouldBeEqualTo PusherState.REGISTERED
+    }
+
+    @Test
+    fun `given a persisted push entity and SetPush API fails when adding Pusher then mutates persisted result with Failed registration state and rethrows error`() {
+        val realmResult = PusherEntity()
+        monarchy.givenWhereReturns(result = realmResult)
+        pushersAPI.givenSetPusherErrors(SocketException())
+
+        assertFailsWith<SocketException> {
+            runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+        }
+
+        realmResult.state shouldBeEqualTo PusherState.FAILED_TO_REGISTER
+    }
+
+    @Test
+    fun `given no persisted push entity and SetPush API fails when adding Pusher then rethrows error`() {
+        monarchy.givenWhereReturns<PusherEntity>(result = null)
+        pushersAPI.givenSetPusherErrors(SocketException())
+
+        assertFailsWith<SocketException> {
+            runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+        }
+    }
+}
+
+internal class FakePushersAPI : PushersAPI {
+
+    private var setRequestPayload: JsonPusher? = null
+    private var error: Throwable? = null
+
+    override suspend fun getPushers(): GetPushersResponse {
+        TODO("Not yet implemented")
```

Comment:
the current tests don't need this yet, hence the throwing TODO, ideally this would be implemented when a test around fetching pushes is created